### PR TITLE
fix(datadog_logs sink): use estimated JSON size for batching and count escapable characters

### PR DESCRIPTION
**Pull request by lukesteensen**
Ref #10020

### CODE REVIEW

File: lib/vector-core/src/event/estimated_json_encoded_size_of.rs [hunk 1]
```diff
@@ -83,17 +83,24 @@ impl EstimatedJsonEncodedSizeOf for Value {
...
+        JsonSize::new(QUOTES_SIZE + self.len() + count_escapes(self.as_bytes()))
     }
 }
 
+// This is an imprecise estimate of the number of characters that will need to be escaped when
+// serializing to JSON. The goal is improve our overall accuracy in the case of relatively common
+// scenarios like strings of nested JSON, etc.
+fn count_escapes(input: &[u8]) -> usize {
+    memchr::memchr3_iter(b'"', b'\\', b'\t', input).count()
+        + memchr::memchr2_iter(b'\n', b'\r', input).count()
```
Comment (bruceg) [inline lib/vector-core/src/event/estimated_json_encoded_size_of.rs:101]: Have you tried doing it in reverse? i.e. counting the number of times that `"\"\\\t\n\r".contains(i)` for each `i` in `input`?

Comment (lukesteensen) [reply lib/vector-core/src/event/estimated_json_encoded_size_of.rs]: In an isolated benchmark, this is much faster for very small strings and much much slower for everything else.

**Pull request closed without merging**